File: plugins/ytqjk-knowledge/skills/ytqjk-knowledge/scripts/migrations.py
"""Transactional v1-v4 SQLite migrations for YTQJK Knowledge."""

from __future__ import annotations

import sqlite3
from datetime import datetime, timezone

from .feedback_migration import downgrade as _downgrade_feedback
from .feedback_migration import repair as _repair_feedback
from .feedback_migration import upgrade as _upgrade_feedback
from .import_migration import DROPS as _IMPORT_DROPS
from .import_migration import repair as _repair_import
from .import_migration import upgrade as _upgrade_import


LATEST_VERSION = 4


def migrate(connection: sqlite3.Connection, target: int = LATEST_VERSION) -> None:
    """Lock, read version, migrate, and repair idempotently in one transaction."""
    if target not in (1, 2, 3, 4):
        raise ValueError("unsupported schema version")
    connection.execute("BEGIN IMMEDIATE")
    try:
        current = int(connection.execute("PRAGMA user_version").fetchone()[0])
        while current < target:
            _upgrade(connection, current + 1)
            current += 1
        while current > target:
            _downgrade(connection, current)
            current -= 1
        if current >= 1:
            _repair_v1(connection, preserve_feedback=current >= 4)
        if current >= 2:
            _execute(connection, _SNAPSHOT_TRIGGER_DROPS + _SNAPSHOT_TRIGGERS)
        if current >= 3:
            _repair_import(connection)
        if current >= 4:
            _repair_feedback(connection)
        connection.execute(f"PRAGMA user_version = {current}")
        connection.commit()
    except BaseException:
        connection.rollback()
        raise


def _upgrade(connection: sqlite3.Connection, version: int) -> None:
    if version == 1:
        _execute(connection, _V1_TABLES)
    elif version == 2:
        _execute(connection, _V2_TABLES)
    elif version == 3:
        _upgrade_import(connection)
    elif version == 4:
        _upgrade_feedback(connection)
    else:
        raise ValueError("unsupported schema version")


def _downgrade(connection: sqlite3.Connection, version: int) -> None:
    if version == 4:
        _downgrade_feedback(connection)
        return
    if version == 3:
        _execute(connection, _IMPORT_DROPS)
        return
    if version != 2:
        raise ValueError("unsupported schema version")
    _execute(connection, _SNAPSHOT_TRIGGER_DROPS + _V2_TABLE_DROPS)


def _repair_v1(
    connection: sqlite3.Connection, *, preserve_feedback: bool = False
) -> None:
    if connection.execute(
        "SELECT 1 FROM sqlite_master WHERE type = 'table' AND name = 'jobs'"
    ).fetchone() is None:
        connection.execute(_JOBS_TABLE)
    columns = {
        str(row[1]) for row in connection.execute("PRAGMA table_info(jobs)")
    }
    for name, definition in _JOB_COLUMNS:
        if name not in columns:
            connection.execute(f"ALTER TABLE jobs ADD COLUMN {name} {definition}")
    _repair_running_jobs(connection)
    drops = _V1_TRIGGER_DROPS
    triggers = _V1_TRIGGERS
    if preserve_feedback:
        protected = "versions_state_machine"
        drops = tuple(item for item in drops if protected not in item)
        triggers = tuple(item for item in triggers if protected not in item)
    _execute(connection, drops + _JOB_TRIGGER_DROPS + triggers + _JOB_TRIGGERS)


def _repair_running_jobs(connection: sqlite3.Connection) -> None:
    reference = datetime.now(timezone.utc)
    recoverable: list[tuple[int]] = []
    live_count = 0
    for job_id, owner, heartbeat, raw_lease in connection.execute(
        "SELECT id, owner, heartbeat_at, lease_expires_at FROM jobs "
        "WHERE state = 'RUNNING' ORDER BY id"
    ):
        if owner is None or heartbeat is None or raw_lease is None:
            recoverable.append((int(job_id),))
            continue
        try:
            lease = datetime.fromisoformat(raw_lease)
        except (TypeError, ValueError) as error:
            raise sqlite3.DatabaseError("invalid RUNNING job lease") from error
        if lease.tzinfo is None:
            raise sqlite3.DatabaseError("invalid RUNNING job lease")
        if lease.astimezone(timezone.utc) <= reference:
            recoverable.append((int(job_id),))
        else:
            live_count += 1
    if live_count > 1:
        raise sqlite3.DatabaseError("multiple live RUNNING job leases")
    _execute(connection, _JOB_TRIGGER_DROPS)
    connection.executemany(
        "UPDATE jobs SET state = 'QUEUED', owner = NULL, heartbeat_at = NULL, "
        "lease_expires_at = NULL WHERE id = ?", recoverable,
    )


def _execute(connection: sqlite3.Connection, statements: tuple[str, ...]) -> None:
    for statement in statements:
        connection.execute(statement)


_V1_TABLES = (
    """CREATE TABLE projects (
        id TEXT PRIMARY KEY, name TEXT NOT NULL, scope TEXT NOT NULL,
        alias TEXT NOT NULL UNIQUE, created_at TEXT NOT NULL
    )""",
    """CREATE TABLE originals (
        sha256 TEXT PRIMARY KEY, content BLOB NOT NULL, created_at TEXT NOT NULL
    )""",
    """CREATE TABLE documents (
        id TEXT PRIMARY KEY, project_id TEXT NOT NULL REFERENCES projects(id),
        title TEXT NOT NULL, deleted_at TEXT
    )""",
    """CREATE TABLE versions (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        document_id TEXT NOT NULL REFERENCES documents(id), ordinal INTEGER NOT NULL,
        state TEXT NOT NULL CHECK(state IN ('candidate','approved','verified','tombstone')),
        original_sha256 TEXT NOT NULL REFERENCES originals(sha256),
        created_at TEXT NOT NULL, UNIQUE(document_id, ordinal)
    )""",
    """CREATE TABLE chunks (
        id TEXT PRIMARY KEY, version_id INTEGER NOT NULL REFERENCES versions(id),
        ordinal INTEGER NOT NULL, content TEXT NOT NULL, UNIQUE(version_id, ordinal)
    )""",
    """CREATE TABLE sources (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        version_id INTEGER NOT NULL REFERENCES versions(id), kind TEXT NOT NULL,
        locator TEXT NOT NULL
    )""",
    """CREATE TABLE governance (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        version_id INTEGER NOT NULL REFERENCES versions(id), action TEXT NOT NULL,
        actor TEXT NOT NULL, created_at TEXT NOT NULL
    )""",
    """CREATE TABLE audit (
        id INTEGER PRIMARY KEY AUTOINCREMENT, event TEXT NOT NULL,
        subject_id TEXT NOT NULL, created_at TEXT NOT NULL, detail TEXT NOT NULL
    )""",
    """CREATE TABLE jobs (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        kind TEXT NOT NULL CHECK(kind IN ('create_project','create_candidate',
            'edit_candidate','soft_delete_candidate','append_state','create_snapshot')),
        payload TEXT NOT NULL, state TEXT NOT NULL CHECK(state IN
            ('QUEUED','RUNNING','SUCCEEDED','FAILED')), dedupe_key TEXT UNIQUE,
        error TEXT, created_at TEXT NOT NULL, started_at TEXT, finished_at TEXT,
        owner TEXT, lease_expires_at TEXT, heartbeat_at TEXT,
        attempt INTEGER NOT NULL DEFAULT 0
    )""",
)

_V2_TABLES = (
    """CREATE TABLE snapshots (
        id TEXT PRIMARY KEY, project_id TEXT NOT NULL REFERENCES projects(id),
        generation INTEGER NOT NULL, state TEXT NOT NULL CHECK(state IN ('BUILDING','ACTIVE')),
        created_at TEXT NOT NULL, UNIQUE(project_id, generation)
    )""",
    """CREATE TABLE snapshot_versions (
        snapshot_id TEXT NOT NULL REFERENCES snapshots(id),
        document_id TEXT NOT NULL REFERENCES documents(id),
        version_id INTEGER NOT NULL REFERENCES versions(id),
        PRIMARY KEY(snapshot_id, document_id)
    )""",
    """CREATE TABLE active_snapshots (
        project_id TEXT PRIMARY KEY REFERENCES projects(id),
        snapshot_id TEXT NOT NULL REFERENCES snapshots(id)
    )""",
)

_JOBS_TABLE = _V1_TABLES[-1]

_JOB_COLUMNS = (
    ("owner", "TEXT"),
    ("lease_expires_at", "TEXT"),
    ("heartbeat_at", "TEXT"),
    ("attempt", "INTEGER NOT NULL DEFAULT 0"),
)

_V1_TRIGGER_DROPS = (
    "DROP TRIGGER IF EXISTS projects_immutable",
    "DROP TRIGGER IF EXISTS documents_soft_delete_candidate",
    "DROP TRIGGER IF EXISTS originals_immutable_update",
    "DROP TRIGGER IF EXISTS originals_immutable_delete",
    "DROP TRIGGER IF EXISTS versions_append_only",
    "DROP TRIGGER IF EXISTS versions_no_delete",
    "DROP TRIGGER IF EXISTS versions_state_machine",
    "DROP TRIGGER IF EXISTS audit_immutable_update",
    "DROP TRIGGER IF EXISTS audit_immutable_delete",
)

_V1_TRIGGERS = (
    """CREATE TRIGGER projects_immutable BEFORE UPDATE OF id, scope, alias ON projects
        BEGIN SELECT RAISE(ABORT, 'project identity is immutable'); END""",
    """CREATE TRIGGER documents_soft_delete_candidate BEFORE UPDATE OF deleted_at ON documents
        WHEN NOT (OLD.deleted_at IS NULL AND NEW.deleted_at IS NOT NULL AND EXISTS (
            SELECT 1 FROM versions WHERE document_id = OLD.id AND ordinal = (
                SELECT MAX(ordinal) FROM versions WHERE document_id = OLD.id)
            AND state = 'candidate'))
        BEGIN SELECT RAISE(ABORT, 'only active candidates can be soft deleted'); END""",
    """CREATE TRIGGER originals_immutable_update BEFORE UPDATE ON originals
        BEGIN SELECT RAISE(ABORT, 'originals are immutable'); END""",
    """CREATE TRIGGER originals_immutable_delete BEFORE DELETE ON originals
        BEGIN SELECT RAISE(ABORT, 'originals are immutable'); END""",
    """CREATE TRIGGER versions_append_only BEFORE UPDATE ON versions
        BEGIN SELECT RAISE(ABORT, 'versions are append-only'); END""",
    """CREATE TRIGGER versions_no_delete BEFORE DELETE ON versions
        BEGIN SELECT RAISE(ABORT, 'versions are append-only'); END""",
    """CREATE TRIGGER versions_state_machine BEFORE INSERT ON versions
        WHEN NEW.ordinal != COALESCE((SELECT MAX(ordinal) + 1 FROM versions
            WHERE document_id = NEW.document_id), 1) OR NOT (
            (NEW.ordinal = 1 AND NEW.state = 'candidate') OR EXISTS (
                SELECT 1 FROM versions prior WHERE prior.document_id = NEW.document_id
                AND prior.ordinal = NEW.ordinal - 1 AND (
                    (prior.state = 'candidate' AND NEW.state IN
                        ('candidate','approved','verified','tombstone')) OR
                    (prior.state = 'approved' AND NEW.state IN ('verified','tombstone')) OR
                    (prior.state = 'verified' AND NEW.state = 'tombstone'))))
        BEGIN SELECT RAISE(ABORT, 'invalid governance state transition'); END""",
    """CREATE TRIGGER audit_immutable_update BEFORE UPDATE ON audit
        BEGIN SELECT RAISE(ABORT, 'audit is append-only'); END""",
    """CREATE TRIGGER audit_immutable_delete BEFORE DELETE ON audit
        BEGIN SELECT RAISE(ABORT, 'audit is append-only'); END""",
)

_JOB_TRIGGER_DROPS = (
    "DROP TRIGGER IF EXISTS jobs_insert_guard",
    "DROP TRIGGER IF EXISTS jobs_payload_immutable",
    "DROP TRIGGER IF EXISTS jobs_state_machine",
    "DROP TRIGGER IF EXISTS jobs_lease_guard",
)

_SNAPSHOT_TRIGGER_DROPS = (
    "DROP TRIGGER IF EXISTS snapshots_insert_guard",
    "DROP TRIGGER IF EXISTS snapshots_immutable",
    "DROP TRIGGER IF EXISTS snapshots_no_delete",
    "DROP TRIGGER IF EXISTS snapshot_versions_insert_guard",
    "DROP TRIGGER IF EXISTS snapshot_versions_immutable",
    "DROP TRIGGER IF EXISTS snapshot_versions_no_delete",
    "DROP TRIGGER IF EXISTS active_snapshots_insert_guard",
    "DROP TRIGGER IF EXISTS active_snapshots_update_guard",
    "DROP TRIGGER IF EXISTS active_snapshots_no_delete",
)

_JOB_TRIGGERS = (
    """CREATE TRIGGER jobs_insert_guard BEFORE INSERT ON jobs WHEN
        NEW.state != 'QUEUED' OR NEW.owner IS NOT NULL OR NEW.lease_expires_at IS NOT NULL
        OR NEW.heartbeat_at IS NOT NULL OR NEW.attempt != 0
        BEGIN SELECT RAISE(ABORT, 'jobs must begin queued'); END""",
    """CREATE TRIGGER jobs_payload_immutable BEFORE UPDATE OF kind, payload,
        dedupe_key, created_at ON jobs BEGIN SELECT RAISE(ABORT, 'job payload is immutable'); END""",
    """CREATE TRIGGER jobs_state_machine BEFORE UPDATE OF state ON jobs WHEN NOT (
        (OLD.state = 'QUEUED' AND NEW.state = 'RUNNING' AND NEW.owner IS NOT NULL
            AND NEW.lease_expires_at IS NOT NULL AND NEW.heartbeat_at IS NOT NULL
            AND NEW.attempt = OLD.attempt + 1) OR
        (OLD.state = 'RUNNING' AND NEW.state IN ('SUCCEEDED','FAILED')
            AND NEW.owner = OLD.owner AND NEW.attempt = OLD.attempt) OR
        (OLD.state = 'RUNNING' AND NEW.state = 'QUEUED' AND NEW.owner IS NULL
            AND NEW.lease_expires_at IS NULL AND NEW.heartbeat_at IS NULL
            AND NEW.attempt = OLD.attempt AND julianday(OLD.lease_expires_at) <= julianday('now')))
        BEGIN SELECT RAISE(ABORT, 'invalid job state transition'); END""",
    """CREATE TRIGGER jobs_lease_guard BEFORE UPDATE OF owner, lease_expires_at,
        heartbeat_at ON jobs WHEN OLD.state = 'RUNNING' AND NEW.state = 'RUNNING'
        AND (NEW.owner != OLD.owner OR NEW.lease_expires_at IS NULL OR NEW.heartbeat_at IS NULL)
        BEGIN SELECT RAISE(ABORT, 'invalid job lease update'); END""",
)

_SNAPSHOT_TRIGGERS = (
    """CREATE TRIGGER snapshots_insert_guard BEFORE INSERT ON snapshots
        WHEN NEW.state != 'BUILDING' BEGIN SELECT RAISE(ABORT, 'snapshots begin building'); END""",
    """CREATE TRIGGER snapshots_immutable BEFORE UPDATE ON snapshots WHEN NOT
        (OLD.state = 'BUILDING' AND NEW.state = 'ACTIVE' AND NEW.id = OLD.id
        AND NEW.project_id = OLD.project_id AND NEW.generation = OLD.generation
        AND NEW.created_at = OLD.created_at)
        BEGIN SELECT RAISE(ABORT, 'snapshots are immutable'); END""",
    """CREATE TRIGGER snapshots_no_delete BEFORE DELETE ON snapshots
        BEGIN SELECT RAISE(ABORT, 'snapshots are immutable'); END""",
    """CREATE TRIGGER snapshot_versions_insert_guard BEFORE INSERT ON snapshot_versions
        WHEN NOT EXISTS (SELECT 1 FROM snapshots s JOIN documents d ON d.id = NEW.document_id
            JOIN versions v ON v.id = NEW.version_id AND v.document_id = d.id
            WHERE s.id = NEW.snapshot_id AND s.project_id = d.project_id AND s.state = 'BUILDING')
        BEGIN SELECT RAISE(ABORT, 'snapshot membership requires building snapshot'); END""",
    """CREATE TRIGGER snapshot_versions_immutable BEFORE UPDATE ON snapshot_versions
        BEGIN SELECT RAISE(ABORT, 'snapshot membership is immutable'); END""",
    """CREATE TRIGGER snapshot_versions_no_delete BEFORE DELETE ON snapshot_versions
        BEGIN SELECT RAISE(ABORT, 'snapshot membership is immutable'); END""",
    """CREATE TRIGGER active_snapshots_insert_guard BEFORE INSERT ON active_snapshots
        WHEN NOT EXISTS (SELECT 1 FROM snapshots WHERE id = NEW.snapshot_id
            AND project_id = NEW.project_id AND state = 'ACTIVE')
        BEGIN SELECT RAISE(ABORT, 'active snapshot must be active generation'); END""",
    """CREATE TRIGGER active_snapshots_update_guard BEFORE UPDATE ON active_snapshots
        WHEN NOT EXISTS (SELECT 1 FROM snapshots WHERE id = NEW.snapshot_id
            AND project_id = NEW.project_id AND state = 'ACTIVE')
        BEGIN SELECT RAISE(ABORT, 'active snapshot must be active generation'); END""",
    """CREATE TRIGGER active_snapshots_no_delete BEFORE DELETE ON active_snapshots
        BEGIN SELECT RAISE(ABORT, 'active snapshot pointer is immutable'); END""",
)

_V2_TABLE_DROPS = (
    "DROP TABLE active_snapshots",
    "DROP TABLE snapshot_versions",
    "DROP TABLE snapshots",
)

File: internal/knowledge/migrations.go
package knowledge

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"strconv"
	"strings"
)

type migrationStep struct {
	version    int
	statements []string
	repair     func(*sql.Conn) error
}

var schemaMigrations = []migrationStep{
	{version: 1, statements: schemaV1Statements, repair: repairJobsSchema},
	{version: 2, statements: schemaV2Statements},
	{version: 3, statements: schemaV3Statements, repair: repairLegacyImportSchema},
	{version: 4, statements: schemaV4Statements},
}

func migrate(database *sql.DB) (string, error) {
	ctx := context.Background()
	connection, err := database.Conn(ctx)
	if err != nil {
		return "", err
	}
	defer connection.Close()
	if _, err := connection.ExecContext(ctx, "BEGIN IMMEDIATE"); err != nil {
		return "", err
	}
	committed := false
	defer func() {
		if !committed {
			_, _ = connection.ExecContext(ctx, "ROLLBACK")
		}
	}()
	var currentVersion int
	if err := connection.QueryRowContext(ctx, "PRAGMA user_version").Scan(&currentVersion); err != nil {
		return "", err
	}
	if currentVersion > LatestSchema {
		return "", fmt.Errorf("knowledge schema %d is newer than supported schema %d", currentVersion, LatestSchema)
	}

	for _, step := range schemaMigrations {
		if err := applyMigrationStep(ctx, connection, step); err != nil {
			return "", err
		}
		if currentVersion < step.version {
			if _, err := connection.ExecContext(ctx, "PRAGMA user_version = "+strconv.Itoa(step.version)); err != nil {
				return "", fmt.Errorf("set knowledge schema v%d: %w", step.version, err)
			}
		}
	}
	if err := ensureJobStateMachineTriggers(ctx, connection); err != nil {
		return "", fmt.Errorf("repair knowledge job state machines: %w", err)
	}
	if err := recoverStaleJobLeases(ctx, connection); err != nil {
		return "", fmt.Errorf("recover knowledge job leases: %w", err)
	}
	feedbackJobs, err := ensureFeedbackEventRoute(ctx, connection)
	if err != nil {
		return "", fmt.Errorf("validate feedback job route: %w", err)
	}
	if _, err := connection.ExecContext(ctx, "COMMIT"); err != nil {
		return "", err
	}
	committed = true
	return feedbackJobs, nil
}

func ensureFeedbackEventRoute(ctx context.Context, connection *sql.Conn) (string, error) {
	route, err := detectFeedbackJobsTableOn(ctx, connection)
	if err != nil {
		return "", err
	}
	var guardExists int
	if err := connection.QueryRowContext(ctx, `SELECT COUNT(*) FROM sqlite_master
WHERE type='trigger' AND name='feedback_events_insert_guard'`).Scan(&guardExists); err != nil {
		return "", err
	}
	if guardExists != 0 {
		if err := probeFeedbackEventRoute(ctx, connection, route); err == nil {
			return route, nil
		}
		if _, err := connection.ExecContext(ctx, "DROP TRIGGER feedback_events_insert_guard"); err != nil {
			return "", err
		}
	}
	if err := createFeedbackEventGuard(ctx, connection, route); err != nil {
		return "", err
	}
	if err := probeFeedbackEventRoute(ctx, connection, route); err != nil {
		return "", err
	}
	return route, nil
}

func createFeedbackEventGuard(ctx context.Context, connection *sql.Conn, route string) error {
	statement := fmt.Sprintf(`CREATE TRIGGER feedback_events_insert_guard
BEFORE INSERT ON feedback_events WHEN NOT EXISTS (
  SELECT 1 FROM %s WHERE id=NEW.job_id AND kind='record_feedback' AND state='RUNNING')
BEGIN SELECT RAISE(ABORT, 'feedback event requires running job'); END`, route)
	_, err := connection.ExecContext(ctx, statement)
	return err
}

func probeFeedbackEventRoute(
	ctx context.Context,
	connection *sql.Conn,
	route string,
) (returnedErr error) {
	if _, err := connection.ExecContext(ctx, "SAVEPOINT feedback_route_probe"); err != nil {
		return err
	}
	defer func() {
		_, rollbackErr := connection.ExecContext(ctx, "ROLLBACK TO feedback_route_probe")
		_, releaseErr := connection.ExecContext(ctx, "RELEASE feedback_route_probe")
		returnedErr = errors.Join(returnedErr, rollbackErr, releaseErr)
	}()
	projectID, err := newUUID()
	if err != nil {
		return err
	}
	documentID, err := newUUID()
	if err != nil {
		return err
	}
	validInvocation, err := newUUID()
	if err != nil {
		return err
	}
	invalidInvocation, err := newUUID()
	if err != nil {
		return err
	}
	nonFeedbackInvocation, err := newUUID()
	if err != nil {
		return err
	}
	originalSHA := strings.TrimPrefix(
		dedupeKey("feedback-route-probe", map[string]any{"project_id": projectID}),
		"feedback-route-probe:",
	)
	statements := []struct {
		query string
		args  []any
	}{
		{"INSERT INTO projects(id,name,scope,alias,created_at) VALUES (?,?,?,?,?)",
			[]any{projectID, "feedback route probe", "project", "probe-" + projectID, "probe"}},
		{"INSERT INTO originals(sha256,content,created_at) VALUES (?,?,?)",
			[]any{originalSHA, "probe", "probe"}},
		{"INSERT INTO documents(id,project_id,title) VALUES (?,?,?)",
			[]any{documentID, projectID, "feedback route probe"}},
	}
	for _, statement := range statements {
		if _, err := connection.ExecContext(ctx, statement.query, statement.args...); err != nil {
			return err
		}
	}
	versionResult, err := connection.ExecContext(ctx, `INSERT INTO versions(
document_id,ordinal,state,original_sha256,created_at) VALUES (?,1,'candidate',?,'probe')`,
		documentID, originalSHA)
	if err != nil {
		return err
	}
	versionID, err := versionResult.LastInsertId()
	if err != nil {
		return err
	}
	var validJobID int64
	if err := connection.QueryRowContext(ctx, `SELECT MIN(
COALESCE((SELECT MIN(id) FROM jobs),0),
COALESCE((SELECT MIN(id) FROM feedback_jobs),0),0)-1`).Scan(&validJobID); err != nil {
		return err
	}
	invalidJobID := validJobID - 1
	for _, job := range []struct {
		id  int64
		key string
	}{{validJobID, "valid-" + projectID}, {invalidJobID, "invalid-" + projectID}} {
		query := "INSERT INTO " + route +
			"(id,kind,payload,state,dedupe_key,created_at) VALUES (?,'record_feedback','{}','QUEUED',?,'probe')"
		if _, err := connection.ExecContext(ctx, query, job.id, job.key); err != nil {
			return fmt.Errorf("feedback job table does not accept record_feedback: %w", err)
		}
	}
	if _, err := connection.ExecContext(ctx, "UPDATE "+route+` SET state='RUNNING',
owner='probe',lease_expires_at='2999-01-01T00:00:00Z',heartbeat_at='probe',attempt=attempt+1
WHERE id=?`, validJobID); err != nil {
		return err
	}
	insertEvent := `INSERT INTO feedback_events(job_id,document_id,invocation_id,correct,score,state,
input_version_id,result_version_id,created_at) VALUES (?,?,?,1,1,'candidate',?,?,'probe')`
	if _, err := connection.ExecContext(
		ctx, insertEvent, validJobID, documentID, validInvocation, versionID, versionID,
	); err != nil {
		return fmt.Errorf("feedback event guard rejected a running job: %w", err)
	}
	if _, err := connection.ExecContext(
		ctx, insertEvent, invalidJobID, documentID, invalidInvocation, versionID, versionID,
	); err == nil {
		return errors.New("feedback event guard accepted a queued job")
	}
	if route == jobsTable {
		nonFeedbackJobID := invalidJobID - 1
		if _, err := connection.ExecContext(ctx, `INSERT INTO jobs(
id,kind,payload,state,dedupe_key,created_at)
VALUES (?,'create_project','{}','QUEUED',?,'probe')`,
			nonFeedbackJobID, "non-feedback-"+projectID); err != nil {
			return err
		}
		if _, err := connection.ExecContext(ctx, `UPDATE jobs SET state='RUNNING',
owner='probe',lease_expires_at='2999-01-01T00:00:00Z',heartbeat_at='probe',attempt=attempt+1
WHERE id=?`, nonFeedbackJobID); err != nil {
			return err
		}
		if _, err := connection.ExecContext(
			ctx, insertEvent, nonFeedbackJobID, documentID, nonFeedbackInvocation, versionID, versionID,
		); err == nil {
			return errors.New("feedback event guard accepted a non-feedback job")
		}
	}
	return nil
}

func ensureJobStateMachineTriggers(ctx context.Context, connection *sql.Conn) error {
	triggers := []struct {
		name       string
		statements []string
	}{
		{name: "jobs_state_machine", statements: schemaV1Statements},
		{name: "feedback_jobs_state_machine", statements: schemaV4Statements},
	}
	for _, trigger := range triggers {
		var existing string
		if err := connection.QueryRowContext(
			ctx, "SELECT sql FROM sqlite_master WHERE type='trigger' AND name=?", trigger.name,
		).Scan(&existing); err != nil {
			return err
		}
		if strings.Contains(existing, "OLD.lease_expires_at IS NULL") &&
			strings.Contains(existing, "OLD.heartbeat_at IS NULL") {
			continue
		}
		canonical := ""
		prefix := "CREATE TRIGGER IF NOT EXISTS " + trigger.name + " "
		for _, statement := range trigger.statements {
			if strings.HasPrefix(statement, prefix) {
				canonical = statement
				break
			}
		}
		if canonical == "" {
			return fmt.Errorf("canonical trigger %s is missing", trigger.name)
		}
		if _, err := connection.ExecContext(ctx, "DROP TRIGGER "+trigger.name); err != nil {
			return err
		}
		if _, err := connection.ExecContext(ctx, canonical); err != nil {
			return err
		}
	}
	return nil
}

func applyMigrationStep(ctx context.Context, connection *sql.Conn, step migrationStep) error {
	for _, statement := range step.statements {
		if _, err := connection.ExecContext(ctx, statement); err != nil {
			return fmt.Errorf("knowledge schema v%d migration: %w", step.version, err)
		}
	}
	if step.repair != nil {
		if err := step.repair(connection); err != nil {
			return fmt.Errorf("knowledge schema v%d additive repair: %w", step.version, err)
		}
	}
	return nil
}

func repairJobsSchema(connection *sql.Conn) error {
	exists, err := objectExists(connection, "table", "jobs")
	if err != nil || !exists {
		return err
	}
	columns, err := connectionColumns(connection, "jobs")
	if err != nil {
		return err
	}
	optional := []struct{ name, definition string }{
		{"dedupe_key", "TEXT"},
		{"error", "TEXT"},
		{"started_at", "TEXT"},
		{"finished_at", "TEXT"},
		{"owner", "TEXT"},
		{"lease_expires_at", "TEXT"},
		{"heartbeat_at", "TEXT"},
		{"attempt", "INTEGER NOT NULL DEFAULT 0"},
	}
	for _, item := range optional {
		if columns[item.name] {
			continue
		}
		statement := "ALTER TABLE jobs ADD COLUMN " + item.name + " " + item.definition
		if _, err := connection.ExecContext(context.Background(), statement); err != nil {
			return fmt.Errorf("add jobs.%s: %w", item.name, err)
		}
	}
	return nil
}

func repairLegacyImportSchema(connection *sql.Conn) error {
	exists, err := objectExists(connection, "table", "import_provenance")
	if err != nil || !exists {
		return err
	}
	columns, err := connectionColumns(connection, "import_provenance")
	if err != nil {
		return err
	}
	if columns["governance_state"] {
		return nil
	}
	_, err = connection.ExecContext(context.Background(), `ALTER TABLE import_provenance
ADD COLUMN governance_state TEXT NOT NULL DEFAULT 'CANDIDATE' CHECK(governance_state='CANDIDATE')`)
	if err != nil {
		return fmt.Errorf("add import_provenance.governance_state: %w", err)
	}
	return nil
}

func objectExists(connection *sql.Conn, kind, name string) (bool, error) {
	var count int
	err := connection.QueryRowContext(
		context.Background(),
		"SELECT COUNT(*) FROM sqlite_master WHERE type=? AND name=?",
		kind,
		name,
	).Scan(&count)
	return count != 0, err
}

func connectionColumns(connection *sql.Conn, table string) (map[string]bool, error) {
	if !validTable(table) {
		return nil, errors.New("unsupported table")
	}
	rows, err := connection.QueryContext(context.Background(), "PRAGMA table_info("+table+")")
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	result := map[string]bool{}
	for rows.Next() {
		var cid, notNull, primaryKey int
		var name, kind string
		var defaultValue any
		if err := rows.Scan(&cid, &name, &kind, &notNull, &defaultValue, &primaryKey); err != nil {
			return nil, err
		}
		result[name] = true
	}
	return result, rows.Err()
}
